app/js/DocArchiveManager.js

```diff
@@ -1,295 +1,183 @@
+    (doc) => archiveDoc(projectId, doc)
+  )
+}
 
-  destroyDoc(project_id, doc_id, callback) {
-    logger.log({ project_id, doc_id }, 'removing doc from mongo and s3')
-    return MongoManager.findDoc(project_id, doc_id, { inS3: 1 }, function (
-      error,
-      doc
-    ) {
-      if (error != null) {
-        return callback(error)
-      }
-      if (doc == null) {
-        return callback(new Errors.NotFoundError('Doc not found in Mongo'))
-      }
-      if (doc.inS3 === true) {
-        return DocArchive._deleteDocFromS3(project_id, doc_id, function (err) {
-          if (err != null) {
-            return err
-          }
-          return MongoManager.destroyDoc(doc_id, callback)
-        })
-      } else {
-        return MongoManager.destroyDoc(doc_id, callback)
-      }
-    })
-  },
+async function archiveDoc(projectId, doc) {
+  logger.log(
+    { project_id: projectId, doc_id: doc._id },
+    'sending doc to persistor'
+  )
+  const key = `${projectId}/${doc._id}`
 
-  _deleteDocFromS3(project_id, doc_id, callback) {
-    let options
-    try {
-      options = DocArchive.buildS3Options(project_id + '/' + doc_id)
-    } catch (e) {
-      return callback(e)
-    }
-    options.json = true
-    return request.del(options, function (err, res, body) {
-      if (err != null || res.statusCode !== 204) {
-        logger.err(
-          { err, res, project_id, doc_id },
-          'something went wrong deleting doc from aws'
-        )
-        return callback(new Error('Error in S3 request'))
-      }
-      return callback()
-    })
-  },
+  if (doc.lines == null) {
+    throw new Error('doc has no lines')
+  }
 
-  _s3DocToMongoDoc(doc, callback) {
-    if (callback == null) {
-      callback = function (error, mongo_doc) {}
-    }
-    const mongo_doc = {}
-    if (doc.schema_v === 1 && doc.lines != null) {
-      mongo_doc.lines = doc.lines
-      if (doc.ranges != null) {
-        mongo_doc.ranges = RangeManager.jsonRangesToMongo(doc.ranges)
-      }
-    } else if (doc instanceof Array) {
-      mongo_doc.lines = doc
-    } else {
-      return callback(new Error("I don't understand the doc format in s3"))
-    }
-    return callback(null, mongo_doc)
-  },
+  const json = JSON.stringify({
+    lines: doc.lines,
+    ranges: doc.ranges,
+    schema_v: 1
+  })
 
-  _mongoDocToS3Doc(doc, callback) {
-    if (callback == null) {
-      callback = function (error, s3_doc) {}
-    }
-    if (doc.lines == null) {
-      return callback(new Error('doc has no lines'))
-    }
-    const json = JSON.stringify({
-      lines: doc.lines,
-      ranges: doc.ranges,
-      schema_v: 1
+  // this should never happen, but protects against memory-corruption errors that
+  // have happened in the past
+  if (json.indexOf('\u0000') > -1) {
+    const error = new Error('null bytes detected')
+    logger.err({ err: error, doc }, error.message)
+    throw error
+  }
+
+  const md5 = crypto.createHash('md5').update(json).digest('hex')
```

Comment:
The reason why we don't want it to work that way is becase the persistor itself doesn't know where the stream is coming from. The persistor logic goes like this:

- Did the client supply an md5 hash? If so, verify that it matches the data on the backend when it is uploaded.
- Otherwise, compute the md5 hash of the data as it passes through and verify that the remote has the same md5 after it is uploaded.

This is subtly different, as if we know the md5 beforehand we can use the S3/GCS native implementation that validates the data server-side against the supplied hash and fail if it doesn't.

If we don't know it in advance, we can still validate it as far as the data is concerned from object-persistor's perspective, but we don't want to consume all of the data first to calculate the size as it might be quite large, so we calculate as we go, verify the upload afterwards and delete it if it's not right. That is a more expensive way of doing it.

The other advantage of being able to supply the md5 in advance is that object-persistor doesn't know where the stream is coming from. It could be passing through something from a downstream http client. If that client can tell us what the md5 is beforehand then we can use it to validate both the upload to the storage engine (S3/GCS) and the upload from the client to object-persistor as the middle-man.

Given that we've already serialised the content to json here and it shouldn't be huge, I think it makes sense to calculate it here instead.